src/System.Management.Automation/utils/tracing/PSSysLogProvider.cs

```diff
@@ -0,0 +1,464 @@
+            if (context == null)
+            {
+                return String.Empty;
+            }
+
+            object logData = context.GetVariableValue(SpecialVariables.PSLogUserDataPath);
+
+            if (logData == null)
+            {
+                return String.Empty;
+            }
+
+            return logData.ToString();
+        }
+
+        /// <summary>
+        /// Appends exception information
+        /// </summary>
+        /// <param name="sb">string builder</param>
+        /// <param name="except">exception</param>
+        internal static void AppendException(StringBuilder sb, Exception except)
+        {
+            sb.AppendLine(StringUtil.Format(EtwLoggingStrings.ErrorRecordMessage, except.Message));
+
+            IContainsErrorRecord ier = except as IContainsErrorRecord;
+
+            if (ier != null)
+            {
+                ErrorRecord er = ier.ErrorRecord;
+
+                if (er != null)
+                {
+                    sb.AppendLine(StringUtil.Format(EtwLoggingStrings.ErrorRecordId, er.FullyQualifiedErrorId));
+
+                    ErrorDetails details = er.ErrorDetails;
+
+                    if (details != null)
+                    {
+                        sb.AppendLine(StringUtil.Format(EtwLoggingStrings.ErrorRecordRecommendedAction, details.RecommendedAction));
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Appends additional information
+        /// </summary>
+        /// <param name="sb">string builder</param>
+        /// <param name="additionalInfo">additional information</param>
+        private static void AppendAdditionalInfo(StringBuilder sb, Dictionary<String, String> additionalInfo)
+        {
+            if (additionalInfo != null)
+            {
+                foreach (KeyValuePair<String, String> value in additionalInfo)
+                {
+                    sb.AppendLine(StringUtil.Format("{0} = {1}", value.Key, value.Value));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets PSLevel from severity
+        /// </summary>
+        /// <param name="severity">error severity</param>
+        /// <returns>PS log level</returns>
+        private static PSLevel GetPSLevelFromSeverity(string severity)
+        {
+            switch (severity)
+            {
+                case "Critical":
+                case "Error":
+                    return PSLevel.Error;
+                case "Warning":
+                    return PSLevel.Warning;
+                default:
+                    return PSLevel.Informational;
+            }
+        }
+
+        /// <summary>
+        /// Converts log context to string
+        /// </summary>
+        /// <param name="context">log context</param>
+        /// <returns>string representation</returns>
+        private static string LogContextToString(LogContext context)
+        {
+            StringBuilder sb = new StringBuilder();
```

Comment:
With so many appends, it feels like this `StringBuilder` should preallocate space to avoid unnecessary reallocations - just guess or take a typical result and just use that.

Reply:
Good point, I'll do an estimate and use it.

Reply:
Fixed